pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -284,10 +285,11 @@ public class BrokerService implements Closeable {
     private final AtomicBoolean blockedDispatcherOnHighUnackedMsgs = new AtomicBoolean(false);
     private final ConcurrentOpenHashSet<PersistentDispatcherMultipleConsumers> blockedDispatchers;
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
-
-    @Getter
     @VisibleForTesting
     private final DelayedDeliveryTrackerFactory delayedDeliveryTrackerFactory;
+    // InMemoryDelayedDeliveryTrackerFactory is for the purpose of
+    // fallback if recover BucketDelayedDeliveryTracker failed.
+    private volatile DelayedDeliveryTrackerFactory fallbackDelayedDeliveryTrackerFactory;
```

Comment:
Why not reuse the original `delayedDeliveryTrackerFactory`?

Reply:
Because DelayedDeliveryTracker is a per-subscription component, maybe subscription-a recovery fail but other subscriptions could have recovery success.
If reuse `delayedDeliveryTrackerFactory` and the first subscription fails to recover, all the rest of the subscriptions on the broker will fallback to `ImMemoryDelayedDeliveryTracker`.

Reply:
Thanks, got it.